tests/layer/test_hinsage.py

```diff
@@ -568,6 +565,100 @@ def test_hinsage_from_generator():
 
     xin, xout = hs.build()
     model = keras.Model(inputs=xin, outputs=xout)
-    actual = model.predict_generator(gen.flow([1, 2]))
-    expected = np.array([[26, 29], [32, 31]], dtype=np.float32)
-    assert actual == pytest.approx(expected)
+
+    batch_feats = list(gen.flow([1, 2]))
+
+    # manually calculate the output of HinSage. All kernels are tensors of 1s
+    # the prediction nodes are type  "A" : "A" nodes only have "B" neighbours, while "B" nodes have both "A" and "B"
+    # neighbours.
+
+    kernel_1_A = np.ones((8, 1))
+    kernel_1_B = np.ones((4, 1))
+
+    kernel_2_A = np.ones((2, 1))
+    kernel_2_B = np.ones((2, 1))
+
+    for i, feats in enumerate(batch_feats):
+        # 1st layer
+        # aggregate for the prediction node
+        layer_1_out = []
+        head = feats[0][0]
+        B_neighs = feats[0][1]
+
+        neighs_trans = (
+            B_neighs.reshape(
```

Comment:
I think we can pull out a function here to use across all the layers. Also, since the kernels are always ones, I think we can also just sum across that axis? It looks to me like that was the intention behind initialising all the weights to 1:
```python
def transform_neighbours(neighs, dim):
    neighs.reshape(
        1, dim, int(neighs.shape[1] / dim), neighs.shape[2]
    ).sum(axis=-1).mean(2)
```
I'm not sure if I've got that 100% correct, but something along those lines

Reply:
And maybe even
```python
def hinsage_layer(head, neighs_by_type):
    head_trans = head.sum(axis=-1)
    neigh_trains = sum(transform_neighbours(neigh, head.shape[1]) for neigh in neighs_by_type) / len(neighs_by_type)
    return np.concatenate([head_trans, neighs_trans], axis=-1)
```

Reply:
nice, yeah that's much simpler!